File: docs/character_pipeline/tools/bootstrap.py
#!/usr/bin/env python3
"""Safe Ubuntu toolchain bootstrap for the omoba character pipeline.

This script intentionally avoids modifying shared Windows/Linux AI tool roots.
It may create and delete a tiny probe file in the shared root to verify write
access. Prepare mode creates user-local Linux virtual environments only.
"""

from __future__ import annotations

import argparse
import json
import os
import shutil
import subprocess
import sys
import venv
from pathlib import Path
from typing import Any


DEFAULT_SHARED_ROOT = Path("/media/damody/新增磁碟區/AI_Pic")
DEFAULT_LINUX_VENV_ROOT = Path("~/.cache/omoba-character-pipeline/venvs").expanduser()
DEFAULT_VENV_NAME = "character-pipeline"
DEFAULT_CACHE_ROOT = Path("~/.cache/omoba-character-pipeline").expanduser()


def run_cmd(args: list[str], timeout: int = 10) -> tuple[int, str]:
    try:
        proc = subprocess.run(
            args,
            check=False,
            stdout=subprocess.PIPE,
            stderr=subprocess.STDOUT,
            text=True,
            timeout=timeout,
        )
        return proc.returncode, proc.stdout.strip()
    except FileNotFoundError:
        return 127, f"command not found: {args[0]}"
    except subprocess.TimeoutExpired:
        return 124, f"command timed out: {' '.join(args)}"


def item(name: str, status: str, kind: str, message: str, **extra: Any) -> dict[str, Any]:
    out: dict[str, Any] = {
        "name": name,
        "status": status,
        "kind": kind,
        "message": message,
    }
    out.update(extra)
    return out


def check_gpu() -> dict[str, Any]:
    code, out = run_cmd(
        [
            "nvidia-smi",
            "--query-gpu=name,driver_version,memory.total",
            "--format=csv,noheader",
        ]
    )
    if code == 0 and out:
        return item("gpu", "ok", "Info", out)
    return item("gpu", "error", "ToolchainError", out or "nvidia-smi unavailable")


def check_python() -> dict[str, Any]:
    return item("python", "ok", "Info", sys.version.split()[0], executable=sys.executable)


def venv_python(venv_path: Path) -> Path:
    return venv_path / "bin" / "python"


def check_torch(python_executable: Path | None = None) -> dict[str, Any]:
    python = str(python_executable or Path(sys.executable))
    code, out = run_cmd(
        [
            python,
            "-c",
            "import torch; print('torch', torch.__version__); print('cuda', torch.cuda.is_available())",
        ]
    )
    if code != 0:
        return item("torch_cuda", "error", "ToolchainError", out, executable=python)
    lines = out.splitlines()
    cuda = len(lines) >= 2 and lines[1].strip() == "cuda True"
    if not cuda:
        return item("torch_cuda", "error", "ToolchainError", out or "torch cuda unavailable", executable=python)
    return item("torch_cuda", "ok", "Info", out, executable=python)


def check_path(path: Path, name: str, required_files: list[str]) -> dict[str, Any]:
    if not path.exists():
        return item(name, "error", "ToolchainError", f"missing path: {path}", path=str(path))
    missing = [rel for rel in required_files if not (path / rel).exists()]
    if missing:
        return item(
            name,
            "error",
            "ToolchainError",
            f"path exists but required files are missing: {', '.join(missing)}",
            path=str(path),
        )
    return item(name, "ok", "Info", f"path available: {path}", path=str(path))


def check_shared_root(root: Path) -> list[dict[str, Any]]:
    results: list[dict[str, Any]] = []
    if not root.exists():
        return [item("shared_root", "error", "SharedRootError", f"missing shared root: {root}")]
    if not root.is_dir():
        return [item("shared_root", "error", "SharedRootError", f"not a directory: {root}")]
    results.append(item("shared_root_read", "ok", "Info", f"readable: {root}", path=str(root)))

    probe = root / f".omoba_character_pipeline_probe_{os.getpid()}.tmp"
    try:
        probe.write_text("probe\n", encoding="utf-8")
        probe.unlink()
        results.append(item("shared_root_write_probe", "ok", "Info", "probe file created and deleted"))
    except OSError as exc:
        results.append(item("shared_root_write_probe", "error", "SharedRootError", str(exc)))
        try:
            if probe.exists():
                probe.unlink()
        except OSError:
            pass
    return results


def check_windows_markers(shared_root: Path) -> list[dict[str, Any]]:
    markers = [
        shared_root / "stable-diffusion-webui" / "venv" / "Scripts" / "python.exe",
        shared_root / "ComfyUI" / "python_embeded" / "python.exe",
    ]
    results: list[dict[str, Any]] = []
    for marker in markers:
        if marker.exists():
            results.append(
                item(
                    "windows_runtime_marker",
                    "warn",
                    "Info",
                    "Windows runtime marker detected; bootstrap must not modify it",
                    path=str(marker),
                )
            )
    return results


def check_executable(
    command: str,
    name: str,
    args: list[str] | None = None,
    fallback: Path | None = None,
) -> dict[str, Any]:
    exe = shutil.which(command)
    if not exe and fallback and fallback.exists():
        exe = str(fallback)
    if not exe:
        return item(name, "error", "ToolchainError", f"{command} not found on PATH")
    code, out = run_cmd([exe] + (args or []), timeout=10)
    if code != 0:
        return item(name, "error", "ToolchainError", out or f"{command} failed", executable=exe)
    return item(name, "ok", "Info", out.splitlines()[0] if out else exe, executable=exe)


def check_venv(venv_path: Path) -> dict[str, Any]:
    python = venv_python(venv_path)
    if not python.exists():
        return item("linux_venv", "error", "ToolchainError", f"venv python missing: {python}", path=str(venv_path))
    code, out = run_cmd([str(python), "--version"])
    if code != 0:
        return item("linux_venv", "error", "ToolchainError", out, path=str(venv_path))
    return item("linux_venv", "ok", "Info", out, path=str(venv_path), executable=str(python))


def prepare_venv(venv_path: Path) -> dict[str, Any]:
    if venv_path.exists() and venv_python(venv_path).exists():
        return item("prepare_venv", "ok", "Info", "venv already exists", path=str(venv_path))
    venv_path.parent.mkdir(parents=True, exist_ok=True)
    builder = venv.EnvBuilder(with_pip=True, clear=False, symlinks=True)
    builder.create(venv_path)
    return item("prepare_venv", "ok", "Info", "created Linux venv", path=str(venv_path))


def prepare(shared_root: Path, linux_venv_root: Path, venv_name: str) -> dict[str, Any]:
    venv_path = linux_venv_root / venv_name
    checks: list[dict[str, Any]] = []
    checks.extend(check_shared_root(shared_root))
    checks.extend(check_windows_markers(shared_root))
    checks.append(prepare_venv(venv_path))
    checks.append(check_venv(venv_path))
    return {
        "schema_version": 1,
        "mode": "prepare",
        "shared_root": str(shared_root),
        "linux_venv_root": str(linux_venv_root),
        "venv_path": str(venv_path),
        "venv_python": str(venv_python(venv_path)),
        "checks": checks,
        "summary": {
            "errors": sum(1 for check in checks if check["status"] == "error"),
            "warnings": sum(1 for check in checks if check["status"] == "warn"),
        },
    }


def find_blender(cache_root: Path) -> Path | None:
    exe = shutil.which("blender")
    if exe:
        return Path(exe)
    candidates = sorted(cache_root.glob("blender*/blender")) + sorted(cache_root.glob("blender/blender*/blender"))
    return candidates[0] if candidates else None


def diagnose(shared_root: Path, linux_venv_root: Path, venv_name: str, cache_root: Path) -> dict[str, Any]:
    comfy_root = shared_root / "ComfyUI" / "ComfyUI"
    webui_root = shared_root / "stable-diffusion-webui"
    venv_path = linux_venv_root / venv_name
    python = venv_python(venv_path) if venv_python(venv_path).exists() else None
    blender = find_blender(cache_root)

    checks: list[dict[str, Any]] = [
        check_gpu(),
        check_python(),
        check_venv(venv_path),
        check_torch(python),
        item(
            "linux_venv_root",
            "ok" if linux_venv_root.parent.exists() else "warn",
            "Info",
            "Linux venvs should be created outside the shared AI root",
            path=str(linux_venv_root),
        ),
    ]
    checks.extend(check_shared_root(shared_root))
    checks.append(check_path(comfy_root, "comfyui_root", ["main.py", "models"]))
    checks.append(check_path(webui_root, "stable_diffusion_webui_root", ["launch.py", "models"]))
    checks.extend(check_windows_markers(shared_root))
    checks.append(check_executable("blender", "blender", ["--version"], fallback=blender))
    checks.append(
        check_executable(
            "lua5.4",
            "lua5.4",
            ["-v"],
            fallback=Path("~/.cache/omoba-character-pipeline/bin/lua5.4").expanduser(),
        )
    )

    return {
        "schema_version": 1,
        "mode": "diagnose",
        "shared_root": str(shared_root),
        "linux_venv_root": str(linux_venv_root),
        "venv_path": str(venv_path),
        "venv_python": str(venv_python(venv_path)),
        "cache_root": str(cache_root),
        "checks": checks,
        "summary": {
            "errors": sum(1 for check in checks if check["status"] == "error"),
            "warnings": sum(1 for check in checks if check["status"] == "warn"),
        },
    }


def main() -> int:
    parser = argparse.ArgumentParser(description="Prepare or diagnose omoba character pipeline toolchain")
    parser.add_argument("--diagnose", action="store_true", help="run safe diagnostics")
    parser.add_argument("--prepare", action="store_true", help="create user-local Linux venv")
    parser.add_argument("--shared-root", default=str(DEFAULT_SHARED_ROOT))
    parser.add_argument("--linux-venv-root", default=str(DEFAULT_LINUX_VENV_ROOT))
    parser.add_argument("--venv-name", default=DEFAULT_VENV_NAME)
    parser.add_argument("--cache-root", default=str(DEFAULT_CACHE_ROOT))
    args = parser.parse_args()

    if args.prepare == args.diagnose:
        parser.error("choose exactly one of --prepare or --diagnose")

    shared_root = Path(args.shared_root)
    linux_venv_root = Path(args.linux_venv_root).expanduser()
    cache_root = Path(args.cache_root).expanduser()
    if args.prepare:
        report = prepare(shared_root, linux_venv_root, args.venv_name)
    else:
        report = diagnose(shared_root, linux_venv_root, args.venv_name, cache_root)
    print(json.dumps(report, ensure_ascii=False, indent=2))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: docs/character_pipeline/tools/bootstrap.lua
local script = debug.getinfo(1, "S").source:sub(2)
local root = script:gsub("[/\\]docs[/\\]character_pipeline[/\\]tools[/\\]bootstrap.lua$", "")
package.path = root .. "/?.lua;" .. package.path

local args = require("tools.lua.lib.args").parse(arg)
local path = require("tools.lua.lib.path")
local process = require("tools.lua.lib.process")
local json = require("tools.lua.lib.json")
local platform = require("tools.lua.lib.platform")

assert((args.prepare and not args.diagnose) or (args.diagnose and not args.prepare),
  "choose exactly one of --prepare or --diagnose")

local home = os.getenv(platform.is_windows and "USERPROFILE" or "HOME") or ""
local function expand(value) return tostring(value):gsub("^~", home) end
local shared = expand(args["shared-root"] or "/media/damody/新增磁碟區/AI_Pic")
local vroot = expand(args["linux-venv-root"] or "~/.cache/omoba-character-pipeline/venvs")
local name = args["venv-name"] or "character-pipeline"
local cache = expand(args["cache-root"] or "~/.cache/omoba-character-pipeline")
local venv = path.join(vroot, name)
local python = path.join(venv, platform.is_windows and "Scripts/python.exe" or "bin/python")
local checks = {}

local function item(check_name, status, kind, message, extra)
  local value = extra or {}
  value.name, value.status, value.kind, value.message = check_name, status, kind, message
  table.insert(checks, value)
  return value
end

local function command(check_name, exe, argv, options)
  options = options or {}
  local ok, result = pcall(process.run, exe, argv, { cwd = options.cwd, check = false })
  if not ok then
    return item(check_name, "warn", "Executable", tostring(result),
      { executable = exe, exit_code = json.null })
  end
  local output = (result.stdout .. result.stderr):gsub("%s+$", "")
  return item(check_name, result.exit_code == 0 and "ok" or "warn", "Executable",
    output ~= "" and output or (result.exit_code == 0 and "completed" or "command failed"),
    { executable = exe, exit_code = result.exit_code })
end

local function first_existing(candidates)
  for _, candidate in ipairs(candidates) do if path.is_file(candidate) then return candidate end end
end

if args.prepare then
  assert(not path.absolute(venv):lower():find(path.absolute(shared):lower(), 1, true),
    "Linux venv must not be inside shared AI root")
  path.mkdir_p(vroot)
  if not path.is_file(python) then
    local result = process.run(platform.is_windows and "py" or "python3",
      { "-m", "venv", venv }, { check = false })
    assert(result.exit_code == 0, "venv creation failed: " .. result.stderr)
  end
  command("venv_python", python, { "--version" })
  print(json.encode({ schema_version = 1, mode = "prepare", shared_root = shared,
    linux_venv_root = vroot, venv_path = venv, venv_python = python, checks = checks }))
  return
end

command("gpu", platform.is_windows and "nvidia-smi.exe" or "nvidia-smi",
  { "--query-gpu=name,driver_version", "--format=csv,noheader" })
command("python", python, { "--version" })
if path.is_file(python) then
  command("torch_cuda", python, { "-c",
    "import torch; print(torch.__version__, torch.cuda.is_available()); raise SystemExit(0 if torch.cuda.is_available() else 2)" })
else
  item("torch_cuda", "warn", "PythonPackage", "venv Python missing", { path = python })
end

local comfy = path.join(shared, "ComfyUI", "ComfyUI")
local webui = path.join(shared, "stable-diffusion-webui")
for _, entry in ipairs({ { "shared_root", shared }, { "comfyui_root", comfy },
  { "stable_diffusion_webui_root", webui } }) do
  local exists = path.is_directory(entry[2])
  item(entry[1], exists and "ok" or "warn", "Path",
    exists and "path available" or "path missing", { path = entry[2] })
end

if path.is_file(python) and path.is_file(path.join(comfy, "folder_paths.py")) then
  command("comfyui_smoke", python,
    { "-c", "import folder_paths; print('comfyui import smoke ok')" }, { cwd = comfy })
else
  item("comfyui_smoke", "warn", "Smoke",
    "ComfyUI root or selected venv Python missing; smoke not run",
    { executable = python, path = comfy, exit_code = json.null })
end

local blender = args.blender or os.getenv("BLENDER_EXE") or first_existing({
  path.join(cache, "blender", "blender.exe"), path.join(cache, "blender", "blender"),
  path.join(cache, "bin", "blender.exe"), path.join(cache, "bin", "blender")
}) or (platform.is_windows and "blender.exe" or "blender")
command("blender_background_smoke", blender,
  { "--background", "--factory-startup", "--python-expr",
    "print('blender background smoke ok')" })

for _, marker in ipairs({
  path.join(shared, "stable-diffusion-webui", "venv", "Scripts", "python.exe"),
  path.join(shared, "ComfyUI", "python_embeded", "python.exe")
}) do
  item("windows_runtime_marker", path.exists(marker) and "ok" or "warn", "Safety",
    "read-only marker; never executed", { path = marker })
end

local errors, warnings = 0, 0
for _, check in ipairs(checks) do
  if check.status == "error" then errors = errors + 1
  elseif check.status == "warn" then warnings = warnings + 1 end
end
print(json.encode({ schema_version = 1, mode = "diagnose", shared_root = shared,
  linux_venv_root = vroot, venv_path = venv, venv_python = python, cache_root = cache,
  checks = checks, summary = { errors = errors, warnings = warnings } }))
